R/knitr-engine.R

```diff
@@ -268,3 +282,52 @@ eng_python_wrap <- function(outputs, options) {
   wrap <- yoink("knitr", "wrap")
   wrap(outputs, options)
 }
+
+save_python_session <- function(cache_path) {
+  dill <- tryCatch(
+    import("dill"),
+    error = function(c) {
+      py_error <- py_last_error()
+      if(py_error$type == "ImportError" && py_error$value == "No module named dill") {
+        "No dill"
+      }})
+  if (dill == "No dill") return()
+
+  py_run_string("globals().pop('r')")
```

Comment:
This probably needs to ensure that the `r` / `R` object is truly the 'shim' R object introduced by `reticulate`, since a user could in theory overwrite it.

Reply:
Thanks for the feedback. I'm not actually sure how to check if the class `R` is the one introduced by `reticulate` or the user. Take a look at my latest commit and see if you think it is sufficient.

Reply:
That looks sufficient to me, but you're entirely right that we should have a way of making this more deterministic. Maybe that class should instead be defined within our `rpytools` module?

Reply:
That's a good idea. Then we could possibly do something like this: https://stackoverflow.com/questions/14570802/python-check-if-object-is-instance-of-any-class-from-a-certain-module

I'll see if I can implement something, although I haven't messed with that side of the package before.